kubernetes-sigs/cluster-api-provider-kubemark
# Allow to specify external cluster for hollow pods

**What this PR does / why we need it**:
This code update adds support for running hollow nodes on external cluster infrastructure, so the kubemark worker plane does not interfere with the target cluster.

The way this is achieved is via an additional setting in kubemark machine template: `kubemarkClusterSecretRef` - which is a reference to a secret containing the external cluster kubeconfig and namespace that will be used for hollow pods deployment. Note, that this setting is optional, if the `kubemarkClusterSecretRef` control is missing, the current behavior will be preserved.

**Which issue this PR fixes**: fixes #20

**Special notes for your reviewer**:
Please see inline comments in this PR for review guidance.

**Testing**:
The following was performed to test this feature:

1. Ensure that the current behavior is preserved:

- run `make test-e2e`
- ensure success

2. Test a scenario when `kubemarkClusterSecretRef` is provided in the template, but the actual secret is missing:

- run `make test-e2e`
- ensure test failure due to a missing secret

3. Test a scenario when `kubemarkClusterSecretRef` is provided in the template, and the respective secret is present, but it does not have kubeconfig value:

- run `make test-e2e`
- ensure test failure due to a missing secret value

4. Test a scenario when `kubemarkClusterSecretRef` is provided in the template, and the respective secret is present, and contains the kubeconfig and namespace values:

- run `make test-e2e` to bring up the tenant cluster control plane (along with 4 kubemark nodes running on the same cluster)
- create a new secret containing kubeconfig pointing to the same cluster, but a different namespace `test`
```
cd /tmp
kubectl create ns test
cp ~/.kube/config /tmp/kubeconfig
vi /tmp/kubeconfig ### <-- Note, when running on kind locally, need to replace apiserver endpoing `https://127.0.0.1:XXXXX` with `https://kubernetes.default.svc`, because `127.0.0.1` does not point to the host on a pod running in kind
echo "test" > /tmp/namespace
kubectl create secret generic kubemark-cluster-kubeconfig --type=Opaque --namespace=capk-system --from-file=kubeconfig --from-file=namespace
```
- add 4 more kubemark pods now running on a different namespace `test` according to `kubemark-cluster-kubeconfig`

Output:
```
$ make test-e2e
...
Tenant cluster is ready!
You can access tenant cluster with the above kubeconfig. (It also can be found in '/tmp/km.kubeconfig')
For CNI, you may apply calico with 'kubectl --kubeconfig /tmp/km.kubeconfig apply -f https://docs.projectcalico.org/v3.20/manifests/calico.yaml'

$ kubectl --kubeconfig /tmp/km.kubeconfig apply -f https://docs.projectcalico.org/v3.20/manifests/calico.yaml
...
poddisruptionbudget.policy/calico-kube-controllers created

$ kubectl get nodes --kubeconfig /tmp/km.kubeconfig                                                                
NAME                        STATUS   ROLES                  AGE     VERSION
km-cp-control-plane-9fgk7   Ready    control-plane,master   2m51s   v1.21.1
km-wl-kubemark-md-0-gjf9x   Ready    <none>                 105s    v1.21.1
km-wl-kubemark-md-0-k8vrj   Ready    <none>                 105s    v1.21.1
km-wl-kubemark-md-0-tsltf   Ready    <none>                 105s    v1.21.1
km-wl-kubemark-md-0-tvvb8   Ready    <none>                 105s    v1.21.1

# Now we will add 4 more kubemark pods running on `test` namespace

$ cd /tmp
$ kubectl create ns test
$ cp ~/.kube/config /tmp/kubeconfig
$ vi /tmp/kubeconfig # update apiserver endpoint in kubeconfig. Note, when running on kind locally, need to replace apiserver endpoing `https://127.0.0.1:XXXXX` with `https://kubernetes.default.svc`, because `127.0.0.1` does not point to the host on a pod running in kind
$ echo "test" > /tmp/namespace
$ kubectl create secret generic kubemark-cluster-kubeconfig --type=Opaque --namespace=capk-system --from-file=kubeconfig --from-file=namespace

$ cd $CAPK_PATH
$ kubectl apply -f hack/tests/resources/kubemark-machine-deployment-external-cluster.yaml 

# Note below, 4 additional kubemark nodes

$ k get nodes --kubeconfig /tmp/km.kubeconfig                                                      
NAME                                 STATUS   ROLES                  AGE   VERSION
km-cp-control-plane-9fgk7            Ready    control-plane,master   12m   v1.21.1
km-wl-kubemark-external-md-0-dh6rd   Ready    <none>                 22s   v1.21.1
km-wl-kubemark-external-md-0-lpnp9   Ready    <none>                 21s   v1.21.1
km-wl-kubemark-external-md-0-r475t   Ready    <none>                 22s   v1.21.1
km-wl-kubemark-external-md-0-sttrb   Ready    <none>                 21s   v1.21.1
km-wl-kubemark-md-0-gjf9x            Ready    <none>                 11m   v1.21.1
km-wl-kubemark-md-0-k8vrj            Ready    <none>                 11m   v1.21.1
km-wl-kubemark-md-0-tsltf            Ready    <none>                 11m   v1.21.1
km-wl-kubemark-md-0-tvvb8            Ready    <none>                 11m   v1.21.1

# Note below, 4 kubemark pods are running in default namespace (from initial deployment with 'make test-e2e') and other 4 kubemark pods are running on `test` namespace

$ kubectl get po -A                                                               
NAMESPACE                           NAME                                                             READY   STATUS    RESTARTS   AGE

...

default                             km-wl-kubemark-md-0-gjf9x                                        1/1     Running   0          11m
default                             km-wl-kubemark-md-0-k8vrj                                        1/1     Running   0          11m
default                             km-wl-kubemark-md-0-tsltf                                        1/1     Running   0          11m
default                             km-wl-kubemark-md-0-tvvb8                                        1/1     Running   0          11m
...
test                                km-wl-kubemark-external-md-0-dh6rd                               1/1     Running   0          18s
test                                km-wl-kubemark-external-md-0-lpnp9                               1/1     Running   0          20s
test                                km-wl-kubemark-external-md-0-r475t                               1/1     Running   0          19s
test                                km-wl-kubemark-external-md-0-sttrb                               1/1     Running   0          20s
```

**Release notes**:
```release-note
Support for external cluster for hollow pods
```
